Refactor course category color strategy

gabrielcaires opened the pull request:
Fixes #5605 

### Changes proposed in this Pull Request
* Refactor the colors management to use CSS variables instead of inline styles
* Fix the incorrect style rendering `class="class="`
* Add support for themes that are using CSS variables on the `theme.json` 
* Remove the probs usage. 
* Remove the fallback classes, simplifying the approach.
* Remove the bunch of attributes required to store colors and custom colors.
* Use a custom save method to generate the style and classes, instead of parsing it on the backend.


### Testing instructions.
- Create a page
- Add a course list with the course categories
- Check if there are colors on the course category background
- Publish
- Check if the same colors are rendered
- Back to editor and apply new color (Text Color and Background Color)
- Publish and check if the rendered version is correct
- Apply font size changes and check if it is rendered correctly
-  Apply line height changes and check if it is rendered correctly

### Extra notes

The overall solution was built with some input from @renatho, it is inspired on the  https://github.com/Automattic/sensei/blob/225bc62a19159c2c1f399177307b800f6890c931/includes/blocks/class-sensei-learner-courses-block.php#L44-L51

The previous approach was using a hacky way to detect the theme colors and the implementation was overcomplex, using 6 attributes to store 2 color values, it was also requiring of to use [no standard WordPress method](https://github.com/Automattic/sensei/pull/5610/files#diff-4bc0728ea61a126a0d836cae0019fa4de0802b8a02958bff95e395f3f734718eL72) to generate the CSS and Style attributes because sometimes we were colors slugs and sometimes we were color values.

The new approach is predefining values to [CSS variables](https://github.com/Automattic/sensei/pull/5610/files#diff-1c9577949ccdd36d3039dc35a3cb9c2f3f8b4e1f452fd442550447c8b3579b72) and[ overriding them when custom values were provided by the user](https://github.com/Automattic/sensei/pull/5610/files#diff-1c9577949ccdd36d3039dc35a3cb9c2f3f8b4e1f452fd442550447c8b3579b72).

I am using a smart tip from @renatho [to get the element color and use it on the element background. ](https://github.com/Automattic/sensei/pull/5610/files#diff-1c9577949ccdd36d3039dc35a3cb9c2f3f8b4e1f452fd442550447c8b3579b72R5)


Now we are only storing color values and using CSS variables to avoid adding inline styles on every category link, reducing the HTML size.

gabrielcaires committed:
Refactor the course category colors to use CSS variables

gabrielcaires committed:
Fix php lint

gabrielcaires committed:
Improve test suite

gabrielcaires committed:
Remove unused override

gabrielcaires committed:
Initialize the wrapper array

renatho commented on assets/blocks/course-categories-block/course-categories.scss:
```diff
@@ -1,33 +1,37 @@
-		white-space: nowrap;
-		margin-bottom: 5px;
+.wp-block-#{$block} {
+	--#{$block}-text-color: #ffffff;
```
Is it always white until the user changes it?
If yes, shouldn't we check the contrast and change it to #000 if white is not readable with the current background?

gabrielcaires replied:
As you can see [here](https://github.com/Automattic/sensei/pull/5610/files#diff-1c9577949ccdd36d3039dc35a3cb9c2f3f8b4e1f452fd442550447c8b3579b72R7)  I updated the strategy to use a default background color, following the same strategy that we are following currently on the component.

renatho commented on assets/blocks/course-categories-block/course-categories-edit.js:
```diff
@@ -44,33 +42,45 @@ export function CourseCategoryEdit( props ) {
+	// by default the root backgroundColor and textColor are overwritten by
+	// Gutenberg withColors HOC.
+	useEffect( () => {
+		setAttributes( {
```
If it's not possible to do it tied to a user event, we should use the `__unstableMarkNextChangeAsNotPersistent` action to avoid problems with the undo in the editor.

Does the same apply to the previous `useEffect`?

renatho commented on includes/blocks/class-sensei-course-categories-block.php:
```diff
@@ -69,24 +59,39 @@ public function render_block( $attributes, $content, WP_Block $block ): string {
 
-		$wrapper_attributes = get_block_wrapper_attributes( array( 'class' => $wrapper_classes ) );
-		$link_attributes    = '<a ' . Sensei_Block_Helpers::render_style_attributes( $wrapper_attributes, $css );
+		list( $wrapper_style ) = Sensei_Block_Helpers::css_variables(
```
I wonder if it would make sense to keep all the presentation part in the frontend, and replace only the dynamic part here. Wouldn't it make it simpler?

gabrielcaires replied:
@renatho Could you please provide more details? I didn't follow the suggestion.

renatho replied:
Hi @gabrielcaires!

I mean to use the `save` from the block registration, containing the needed classes and styles (reusing from the `edit`?). And then use this as the replacement (this would come in the `$content` of this method). But please, take into account that I didn't dig into that, so maybe there's a strong reason for this approach that I didn't realize. 😉

gabrielcaires replied:
@renatho sorry for the delay.
I liked your suggestion, so I moved every style and class logic to be done directly on the editor and the PHP size is only getting it and wrapping the list of terms.


My favorites pros are:
- Avoid implementing the rendering part two times, so every time that a new attribute is added is not necessary to update the PHP file. 
-  I could write tests to cover the saving method, giving me more control over it. 
- I  tested it enabling more features using the Gutenberg support and all of them are working fine with no changes


The main point that I haven't a success was removing these lines: 

https://github.com/Automattic/sensei/pull/5610/files#diff-ca728889b5f81916884b66ff250e343f50821f49be6ae2730d2b7dabedf2c45cR61-R68


Unfortunately, the save method doesn't receive the colors from the `withColors`  hook, so I had to save the colors in order to get them on the save method.

renatho replied:
Nice! Thank you for that!
I just couldn't see what's in the link you sent because it didn't work.

gabrielcaires replied:
Oh Sorry, link update here:
https://github.com/Automattic/sensei/pull/5610/files#diff-ca728889b5f81916884b66ff250e343f50821f49be6ae2730d2b7dabedf2c45cR52-R64

I mentioned the useEffects callbacks we are using to update the block attributes.

renatho reviewed (commented):
Hi there! I didn't go deeper in the review and I didn't test it, but in general it looks good to me! I just added a few suggestions. 😉

gabrielcaires committed:
Remove the option color update from the editor undo stack.

gabrielcaires committed:
Disable the contrast checker when the theme is using CSS variables as theme color values

gabrielcaires committed:
Remove unused block attributes

gabrielcaires committed:
Use the correct alignment options

Previously it was using the block alignment `support align` option to control the course-categories text alignment.

gabrielcaires committed:
Move the style/classes rendering to the block editor, instead of do it on php.

gabrielcaires committed:
Merge branch 'trunk' into update/improve-course-category-color-strategy

gabrielcaires committed:
Fix react warning

gabrielcaires committed:
Fix warning when renders a block

gabrielcaires committed:
Save the options attribute

gabrielcaires committed:
Fix lint issues

gabrielcaires committed:
Add missing file

renatho reviewed (approved):
I didn't test this PR. It would be nice if someone with more context on this block could do it. But code looks good to me. :)

LavaGolem reviewed (commented):
Testing colors in Divi (there we use default color)
| Editor | View |
| -- | -- |
| <img width="1233" alt="image" src="https://user-images.githubusercontent.com/7208249/193043765-2e282a41-d8fc-4dfa-8452-08150aa1b89e.png"> | <img width="845" alt="image" src="https://user-images.githubusercontent.com/7208249/193043853-0583bc10-562a-49dc-b262-51771e09f0e4.png"> |

LavaGolem reviewed (approved):
Tested in Astra, Theme color was offered as a color for the course categories
<img width="1428" alt="image" src="https://user-images.githubusercontent.com/7208249/193055223-ca5c6303-552c-45f1-b09b-bdd40cd59a53.png"> 

Displayed in edit and preview when chosen 
<img width="1056" alt="image" src="https://user-images.githubusercontent.com/7208249/193055316-b3888704-efde-4256-bf55-1c1cef346728.png">
